cirq-core/cirq/transformers/measurement_transformers.py

```diff
@@ -0,0 +1,191 @@
+            conditions based on multi-qubit measurements exist. (The latter of
+            these is planned to be implemented soon).
+    """
+
+    context1 = context or transformer_api.TransformerContext()
+    circuit = transformer_primitives.unroll_circuit_op(circuit, deep=True, tags_to_check=None)
+    qubits_found: Set['cirq.Qid'] = set()
+    terminal_measurements: Set['cirq.MeasurementKey'] = set()
+    control_keys: Set['cirq.MeasurementKey'] = set()
+    for op in reversed(list(circuit.all_operations())):
+        gate = op.gate
+        if isinstance(gate, ops.MeasurementGate):
+            key = value.MeasurementKey.parse_serialized(gate.key)
+            if key not in control_keys and qubits_found.isdisjoint(op.qubits):
+                terminal_measurements.add(key)
+        for c in op.classical_controls:
+            control_keys.update(c.keys)
+        qubits_found.update(op.qubits)
+    measurement_qubits: Dict['cirq.MeasurementKey', List['_MeasurementQid']] = {}
+
+    def defer(op: 'cirq.Operation', _) -> 'cirq.OP_TREE':
+        if any(t in context1.ignore_tags for t in op.tags):
+            return op
+        gate = op.gate
+        if isinstance(gate, ops.MeasurementGate):
+            key = value.MeasurementKey.parse_serialized(gate.key)
+            if key in terminal_measurements:
+                return op
+            targets = [_MeasurementQid(key, q) for q in op.qubits]
+            measurement_qubits[key] = targets
+            cxs = [ops.CX(q, target) for q, target in zip(op.qubits, targets)]
+            xs = [ops.X(targets[i]) for i, b in enumerate(gate.full_invert_mask()) if b]
+            return cxs + xs
+        elif protocols.is_measurement(op):
+            return [defer(op, None) for op in protocols.decompose_once(op)]
+        elif op.classical_controls:
+            controls = []
+            for c in op.classical_controls:
+                if isinstance(c, value.KeyCondition):
+                    if c.key not in measurement_qubits:
+                        raise ValueError(f'Deferred measurement for key={c.key} not found.')
+                    qubits = measurement_qubits[c.key]
+                    if len(qubits) != 1:
+                        # TODO: Multi-qubit conditions require
+                        # https://github.com/quantumlib/Cirq/issues/4512
+                        # Remember to update docstring above once this works.
+                        raise ValueError('Only single qubit conditions are allowed.')
+                    controls.extend(qubits)
+                else:
+                    raise ValueError('Only KeyConditions are allowed.')
+            return op.without_classical_controls().controlled_by(
+                *controls, control_values=[tuple(range(1, q.dimension)) for q in controls]
+            )
+        return op
+
+    circuit = transformer_primitives.map_operations(
+        circuit, defer, raise_if_add_qubits=False
+    ).unfreeze()
```

Comment:
Use `transformer_primitives.map_operations_and_unroll` to unroll by default in case of pauli measurements + decompose